sdk/metric/sdk.go

```diff
@@ -0,0 +1,469 @@
+		// referring to this record.  active handles prevent
+		// removing the record from the current map.
+		refcount int64
+
+		// collectedEpoch is the epoch number for which this
+		// record has been exported.  This is modified by the
+		// `Collect()` method.
+		collectedEpoch int64
+
+		// modifiedEpoch is the latest epoch number for which
+		// this record was updated.  Generally, if
+		// modifiedEpoch is less than collectedEpoch, this
+		// record is due for reclaimation.
+		modifiedEpoch int64
+
+		// reclaim is an atomic to control the start of reclaiming.
+		reclaim int64
+
+		// recorder implements the actual RecordOne() API,
+		// depending on the type of aggregation.  If nil, the
+		// metric was disabled by the exporter.
+		recorder export.MetricAggregator
+
+		// next contains the next pointer for both the primary
+		// and the reclaim lists.
+		next doublePtr
+	}
+
+	// singlePointer wraps an unsafe.Pointer and supports basic
+	// load(), store(), clear(), and swapNil() operations.
+	singlePtr struct {
+		ptr unsafe.Pointer
+	}
+
+	// doublePtr is used for the head and next links of two lists.
+	doublePtr struct {
+		primary singlePtr
+		reclaim singlePtr
+	}
+)
+
+var (
+	_ api.Meter           = &SDK{}
+	_ api.LabelSet        = &labels{}
+	_ api.InstrumentImpl  = &instrument{}
+	_ api.HandleImpl      = &record{}
+	_ export.MetricRecord = &record{}
+
+	// hazardRecord is used as a pointer value that indicates the
+	// value is not included in any list.  (`nil` would be
+	// ambiguous, since the final element in a list has `nil` as
+	// the next pointer).
+	hazardRecord = &record{}
+)
+
+func (i *instrument) Meter() api.Meter {
+	return i.meter
+}
+
+func (i *instrument) acquireHandle(ls *labels) *record {
+	// Create lookup key for sync.Map
+	mk := mapkey{
+		descriptor: i.descriptor,
+		encoded:    ls.encoded,
+	}
+
+	// There's a memory allocation here.
+	rec := &record{
+		labels:         ls,
+		descriptor:     i.descriptor,
+		refcount:       1,
+		collectedEpoch: -1,
+		modifiedEpoch:  0,
+	}
+
+	// Load/Store: there's a memory allocation to place `mk` into
+	// an interface here.
+	if actual, loaded := i.meter.current.LoadOrStore(mk, rec); loaded {
+		// Existing record case.
+		rec = actual.(*record)
+		atomic.AddInt64(&rec.refcount, 1)
+		return rec
+	}
+	rec.recorder = i.meter.exporter.AggregatorFor(rec)
+
+	i.meter.addPrimary(rec)
+	return rec
+}
+
+func (i *instrument) AcquireHandle(ls api.LabelSet) api.HandleImpl {
+	labs := i.meter.labsFor(ls)
+	return i.acquireHandle(labs)
+}
+
+func (i *instrument) RecordOne(ctx context.Context, number core.Number, ls api.LabelSet) {
+	ourLs := i.meter.labsFor(ls)
+	h := i.acquireHandle(ourLs)
+	defer h.Release()
+	h.RecordOne(ctx, number)
+}
+
+// New constructs a new *SDK that implements intermediate storage for
+// metric instrument events, suporting configurable aggregation.
+//
+// The SDK does not start any background process to periodically
```

Comment:
That paragraph should rather mention that collecting is controlled by the exporter.